### ISSUE TITLE
[WP1][Enhancement] Feature: L-Module Only HRM Variant (Simplified Architecture)

### ISSUE THREAD

**Issue by KyryloKozlovskyi** [fionntmcc/cross-platform-hrm#39]
## Summary
Implement an L-module only variant of the HRM that removes the Planner (H-module) and outer loop, based on findings from Ge et al. (2025) "Hierarchical Reasoning Models: Perspectives and Misconceptions" (arXiv:2510.00355v2).

### Key Research Findings
The Ge et al. paper from MIT's Center for Brains, Minds, and Machines found that:

1. **H-module is not essential**: An 8-layer L-module only HRM performs similarly to the original 4-layer L + 4-layer H HRM
2. **Training is 2.4x faster**: Runtime dropped from 4h 21m to 1h 48m on A100 GPU
3. **Comparable accuracy**: The simplified model achieves similar or slightly better performance
4. **Confirmed by ARC Foundation**: Independent analysis reached the same conclusion

### Implications for Our Project
This suggests:
- The hierarchical structure may be unnecessary for 4x4 Sudoku
- Computational depth (via iteration) is the key factor, not the specific H/L architecture
- Simpler models are faster to train and easier to deploy on Raspberry Pi 5
- One-step gradient approximation is effectively equivalent to diffusion model training

### Architecture Change
```
ORIGINAL (Current Implementation):
  Input → [Outer Loop (K cycles): [Inner Loop (T steps): Worker] → Planner] → Output
  Total: K × T iterations with Planner updates

SIMPLIFIED (L-Module Only):
  Input → [Single Loop (N steps): Worker] → Output
  Total: N iterations (where N ≈ K × T for equivalence)
```

#### New Model Class: `HRM_4x4_Simple`
```python
class HRM_4x4_Simple(nn.Module):
    """
    Simplified HRM using only the L-module (Worker) without hierarchical structure.
    
    Based on Ge et al. (2025) findings that H-module provides minimal benefit.
    Trains 2.4x faster while maintaining comparable accuracy.
    """
    
    def __init__(
        self,
        hidden_dim: int = 64,
        n_iterations: int = 50,  # Equivalent to K×T in full HRM
        convergence_threshold: float = 1e-3,
        n_worker_layers: int = 8,  # Expanded from 4 layers
        **kwargs
    ):
        ...
```

#### Key Differences from Full HRM
| Aspect | Full HRM | L-Module Only |
|--------|----------|---------------|
| Modules | Worker + Planner | Worker only |
| Hidden states | h_L and h_H | h_L only |
| Iteration structure | K outer × T inner | N flat iterations |
| Parameters | ~117K (4+4 layers) | ~150K (8 layers) |
| Training time | ~4h 21m | ~1h 48m |
| Memory | O(K) outer states | O(1) constant |

### Iteration Logic
```python
def forward(self, x):
    x_in = self.input_net(x)
    h_L = self.h_L_init.expand(batch_size, -1)
    
    # No h_H - Worker iterates with only x_in as context
    for step in range(self.n_iterations):
        h_L_new = self.worker(h_L, context=x_in)
        
        # Check convergence
        if compute_residual(h_L_new, h_L) < self.threshold:
            break
        h_L = h_L_new
    
    return self.output_net(h_L)
```

### Functional Requirements
- Implement `HRM_4x4_Simple` class without Planner module
- Support configurable number of iterations (default: 50)
- Support 8-layer Worker module for expanded capacity
- Maintain same Input/Output network structure
- Return execution trace with iteration count and residuals

### Testing Requirements
- Unit tests for `HRM_4x4_Simple`
- Forward/backward pass verification
- Convergence behavior testing

### Performance Requirements
- Training speed improvement ≥ 2x compared to full HRM
- Validation accuracy within 5% of full HRM baseline
- Memory usage reduction (no h_H state storage)

## Core Implementation
1. Create `hrm/model_simple.py` with `HRM_4x4_Simple`
2. Modify `WorkerModule` to support 8-layer configuration
3. Implement flat iteration loop (no outer/inner structure)
4. Add convergence-based early stopping

### Worker Module Modification
The Worker currently expects `h_H` as context. For L-module only:
```python
# Option 1: Pass x_in as context instead of h_H
h_L_new = worker(h_L, h_H=x_in, x_in=x_in)  # Use x_in for both

# Option 2: Modify Worker to not require h_H
h_L_new = worker(h_L, x_in=x_in)  # Remove h_H dependency
```

### One-Step Gradient Approximation
Still applies - we detach previous h_L:
```python
h_L_prev = h_L.detach()  # No gradient through previous iteration
h_L = worker(h_L_prev, x_in)  # Gradient only through current step
```

### Relation to Diffusion Models
Per Ge et al., this one-step training is equivalent to Latent Consistency Model (LCM) training:
- Model learns to map intermediate state → final answer
- No BPTT required
- O(1) memory regardless of iteration count
